src/Dictionaries/TrieDictionary.cpp

```diff
@@ -331,13 +480,136 @@ void TrieDictionary::loadData()
+
+                UInt8 prefix = std::stoi(addr_str.substr(pos + 1), nullptr, 10);
+
+                addr = addr & IPAddress(prefix, addr.family());
+                ip_records.emplace_back(addr, prefix, row_number);
+            }
+            else
+            {
+                IPAddress addr(addr_str);
+                has_ipv6 = has_ipv6 || (addr.family() == Poco::Net::IPAddress::IPv6);
+                UInt8 prefix = addr.length() * 8;
+                ip_records.emplace_back(addr, prefix, row_number);
             }
         }
     }
 
     stream->readSuffix();
 
+    if (has_ipv6)
+    {
+        auto deleted_count = sort_and_unique(ip_records,
+            [](const auto & record_a, const auto & record_b)
+            {
+                uint8_t a_buf[IPV6_BINARY_LENGTH];
+                uint8_t b_buf[IPV6_BINARY_LENGTH];
+
+                auto cmpres = memcmp16(record_a.asIPv6Binary(a_buf), record_b.asIPv6Binary(b_buf));
+
+                if (cmpres == 0)
+                    return compareTo(record_a.prefixIPv6(), record_b.prefixIPv6());
+                return cmpres;
+            });
+        if (deleted_count > 0)
+            LOG_WARNING(logger, "removing {} non-unique subnets from input", deleted_count);
+
+        auto & ipv6_col = ip_column.emplace<IPv6Container>();
+        ipv6_col.resize_fill(IPV6_BINARY_LENGTH * ip_records.size());
+
+        for (const auto & record : ip_records)
+        {
+            auto ip_array = IPv6ToBinary(record.addr);
+
+            size_t i = row_idx.size();
+            memcpySmallAllowReadWriteOverflow15(&ipv6_col[i * IPV6_BINARY_LENGTH],
+                                                reinterpret_cast<const uint8_t *>(ip_array.data()),
+                                                IPV6_BINARY_LENGTH);
+
+            mask_column.push_back(record.prefixIPv6());
+            row_idx.push_back(record.row);
+        }
+    }
+    else
+    {
+        auto deleted_count = sort_and_unique(ip_records,
+            [](const auto & record_a, const auto & record_b)
+            {
+                UInt32 a = IPv4AsUInt32(record_a.addr.addr());
+                UInt32 b = IPv4AsUInt32(record_b.addr.addr());
+
+                if (a == b)
+                    return compareTo(record_a.prefix, record_b.prefix);
+                return compareTo(a, b);
+            });
+        if (deleted_count > 0)
+            LOG_WARNING(logger, "removing {} non-unique subnets from input", deleted_count);
+
+        auto & ipv4_col = ip_column.emplace<IPv4Container>();
+        ipv4_col.reserve(ip_records.size());
+        for (const auto & record : ip_records)
+        {
+            auto addr = IPv4AsUInt32(record.addr.addr());
+            ipv4_col.push_back(addr);
+            mask_column.push_back(record.prefix);
+            row_idx.push_back(record.row);
+        }
+    }
+
+    parent_subnet.resize(ip_records.size());
+    std::stack<size_t> subnets_stack;
```

Comment:
I think we don't even need stack here.
Because it stores values like `i - 1, parent_subnet[i - 1], parent_subnet[parent_subnet[i - 1]], ...` which are accessible all the time.
But this implementation is also ok.